src/liballoc/collections/btree/map.rs

```diff
@@ -250,8 +250,7 @@ impl<K: Clone + Ord, V: Clone> BTreeClone for BTreeMap<K, V> {
         // the BTree invariants are maintained at the end of the loop
         while !siter.is_empty() {
             if let Some((ok, ov)) = oiter.next() {
-                // SAFETY: This is safe because the `siter.front != siter.back` check
-                // ensures that `siter` is nonempty
+                // SAFETY: This is safe because `siter` is nonempty
```

Comment:
Why is there a safety comment being changed in a PR that's labelled to be about tests?

Reply:
It's something I noticed that wasn't adapted in the history of the already merged change, related to most of the test tweaks. I can do that commit in a separate PR or forget about it.

Reply:
I am not deeply familiar with the code, and unfortunately `next_unchecked` doesn't have a comment explaining its unsafe precondition. But the change makes sense to me, so we can keep it. Just please add a trailing `.` :)

Reply:
Gladly, but to this one only or to all similar comments in the file?

Reply:
Having a trailing `.` after a proper sentence is the recommended comment style in rustc (AFAIK...), but there's no need to increase the diff to port everything. Just, whatever you touch should follow the style.

OTOH if you *want* to add a few more trailing `.`, it's not like that's hard to review.^^

Reply:
> `next_unchecked` doesn't have a comment explaining its unsafe precondition

True. But when I work my way up from the recently documented underlying implementation, the precondition is unhelpful or convoluted. Calling `Range::next_unchecked` is safe if (besides another thing) whoever initialized the range instance, gave it "good" data. Or constructing a valid `Range` is an unsafe operation itself but I don't find information on marking constructors unsafe. So, short unhelpful comment coming in.

Reply:
Well I was trying to review the SAFETY comment and that is hard when the precondition is unclear. ;) But that is not your fault, and your comment doesn't make it worse, so I am still fine with the diff here.

> Or constructing a valid Range is an unsafe operation itself but I don't find information on marking constructors unsafe. 

There is no such thing; we make them private. A constructor can never by itself cause UB.

Reply:
I was adding comments on the `next_unchecked` touched here, because it seems a waste to keep the underlying documentation work hidden.

`Range::next_unchecked` is private too, but we still mark it unsafe (which I found surprising,). Okay, the constructor cannot cause UB itself, but there's an implicit "class invariant" on `Range` (and many others), i.e. `front` and `back` are leaf edges (that's in their type already), point into the same tree (duh...) and `front` is not right of `back` (obviously?). If we don't specify how the constructor should be used, can we silently assume that the class invariant is honoured in `Range::next_unchecked`?

Reply:
Ideally, the invariant should be documented at the type, and everything acting on its fields directly should have `SAFETY` comments explaining why this preserves the invariant. For unsafe methods, well it's up to them whether they require the invariant to already hold or not.

But the code here isn't yet up to that kind of standard. Also if this becomes more documentation changes I would ask you to split the PR, because I don't feel comfortable to review those.

Reply:
> Ideally, the invariant should be documented at the type

Where at the type? For a public type like `Range`, with private fields and safe methods, the usual struct documentation seem a bad place to bother people with internal invariants. `string/vec::from_raw_parts` and `vec::set_len` also list invariants (actually, they are the only places in string and vec listing them), but these are unsafe methods. Finally I found `pub struct FreeRegionMap` that features an invariant in the comments of a field.